src/app/services/myinfo/myinfo.service.ts

```diff
@@ -0,0 +1,315 @@
+   * @param params.singpassEserviceId The eservice id of the form requesting the data.
+   * @returns the person object retrieved.
+   * @throws an error on fetch failure or if circuit breaker is in the opened state. Use {@link CircuitBreaker#isOurError} to determine if a rejection was a result of the circuit breaker or the action.
+   */
+  fetchMyInfoPersonData(
+    params: IPersonBasicRequest,
+  ): ResultAsync<IPersonBasic, CircuitBreakerError | FetchMyInfoError> {
+    return ResultAsync.fromPromise(
+      this.#myInfoClientBreaker.fire(params),
+      (error) => {
+        const logMeta = {
+          action: 'fetchMyInfoPersonData',
+          requestedAttributes: params.requestedAttributes,
+          eServiceId: params.singpassEserviceId,
+        }
+        if (CircuitBreaker.isOurError(error)) {
+          logger.error({
+            message: 'Circuit breaker tripped',
+            meta: logMeta,
+            error,
+          })
+          return new CircuitBreakerError()
+        } else {
+          logger.error({
+            message: 'Error retrieving data from MyInfo',
+            meta: logMeta,
+            error,
+          })
+          return new FetchMyInfoError()
+        }
+      },
+    )
+  }
+
+  /**
+   * Prefill given current form fields with given MyInfo data.
+   * @param myInfoData
+   * @param currFormFields
+   * @returns currFormFields with the MyInfo fields prefilled with data from myInfoData
+   */
+  prefillMyInfoFields(
+    myInfoData: IPersonBasic,
+    currFormFields: IFieldSchema[],
+  ): Result<IPossiblyPrefilledField[], never> {
+    return ok(
+      currFormFields.map((field) => {
+        if (!field?.myInfo?.attr) return field
+
+        const myInfoAttr = field.myInfo.attr
+        const myInfoValue = getMyInfoValue(myInfoAttr, myInfoData)
+        const isReadOnly = isFieldReadOnly(myInfoAttr, myInfoValue, myInfoData)
+        const prefilledField = cloneDeep(field) as IPossiblyPrefilledField
+        prefilledField.fieldValue = myInfoValue
+
+        // Disable field
+        prefilledField.disabled = isReadOnly
+        return prefilledField
+      }),
+    )
+  }
+
+  /**
+   * Saves hashed prefilled values of MyInfo fields.
+   * @param uinFin NRIC
+   * @param formId ID of form being populated
+   * @param prefilledFormFields Fields with fieldValue prefilled and disabled set to true if read-only
+   * @returns the document saved to the database which contains the hashes, or null if the document was not found
+   * @throws error if an error occurred while hashing the values or updating the database
+   */
+  saveMyInfoHashes(
+    uinFin: string,
+    formId: string,
+    prefilledFormFields: IPossiblyPrefilledField[],
+  ): ResultAsync<IMyInfoHashSchema | null, HashingError | DatabaseError> {
+    const readOnlyHashPromises = hashFieldValues(prefilledFormFields)
+    return ResultAsync.fromPromise(
+      Bluebird.props<IHashes>(readOnlyHashPromises),
```

Comment:
This is the only place we use Bluebird. Is there any reason? Should we just use normal promises and then we can delete the Bluebird dependency

Reply:
yeah in fact I added another call to `Bluebird.props`, so now there are 2 instead of 1. I just think it makes a lot more sense than something like `Promise.all` because we care about mapping each attribute to a `Promise<boolean>` and knowing which ones resolve to true/false. If we use `Promise.all` then we have to do something clunky like `_.zip` to line up the attributes and the Promises, which is really unnecessary since we should be using a Map-like data structure in the first place.